### PR TITLE
[impl-senior] moltzap: SDK event decoding at inbound boundary

### LINKED ISSUE #132: [safer:implement-senior] moltzap: SDK event decoding at inbound boundary
Parent epic: #126. Follow-up to #128 (bridge modules).

## Context

From sbd#108 architect design P2: "The follow-up PR must decode raw `@moltzap/app-sdk` events through a schema before passing `MoltzapInbound` into `bridge.onInbound`." The staff-tier bridge PR (#128) locks the interface but treats inbound events as opaque — correct for interface-locking but means the plugin currently accepts any shape from the SDK.

## Acceptance

- Schema validation of raw SDK events at the listener boundary.
- Typed error for malformed events (dropped + diagnostic, no throw).
- Unit tests for valid/invalid/edge shapes.

## Blockers

Waits on #128 merge (interface lives there).

## Non-goals

- SDK protocol changes.
- Buffering malformed events (I7 still applies).

## Linear-project: Moltzap migration

### PR DESCRIPTION
**Pull request by chughtapan**
Closes #132
Architect plan: https://github.com/chughtapan/safer-by-default/issues/108#issuecomment-4275100681

## What changed

Adds schema validation at the listener-to-bridge handoff. `register()` now wraps the user callback with a decode layer: the `MoltzapRegistrar` passes raw `unknown` SDK events; `decodeMoltzapInbound` validates each required field and returns a typed `DecodeError` on failure. Malformed events are reported to the injected `DecodeErrorSink` and dropped with no buffering (I7). The bridge callback only receives well-typed `MoltzapInbound` values.

## Plan anchors

| Change | Plan anchor |
|---|---|
| `DecodeError` type | sbd#108 follow-up: "typed error for malformed events" |
| `DecodeErrorSink` type | sbd#108 follow-up: "dropped + diagnostic" |
| `MoltzapRegistrar` callback `(event: unknown) => void` | sbd#108 follow-up: "decode raw SDK events" |
| `decodeMoltzapInbound()` + `checkField()` internals | sbd#108 follow-up: "through a schema" |
| `register()` `diag` param + `wrappedCb` | sbd#108 follow-up: "before passing MoltzapInbound into bridge.onInbound" |
| 9 new decode-layer tests | Issue #132 Acceptance: "Unit tests for valid/invalid/edge shapes" |

## Scope

- Modules touched: `v2/moltzap/listener.ts`, `test/v2-moltzap-listener.test.ts`
- Tier (from safer-diff-scope): senior — 2 files, internals only, no new module, no new public surface beyond plan-named types
- New modules: 0
- New public signatures outside the plan: 0 (`DecodeError`, `DecodeErrorSink` are plan-named)
- New deps: 0

## Tests

- 7 existing tests updated: added `noopDiag` parameter to all `register()` calls
- 4 new valid-event tests: well-formed decode, extra fields ignored, empty-string messageId, receivedAtMs=0
- 5 new invalid-event tests: null input, missing field, wrong type, array input, multi-event (valid then invalid)

## Simplify

- Applied: extracted `checkField()` helper to de-duplicate the 5 field-validation blocks
- Applied: trimmed "what" comment on decoder to forward-compat WHY only
- Skip: shared decoder utility with `v2/gateway.ts#decodeIssueCommentPayload` — extracting a shared module is architect-tier (new module); noted for future architect pass

## Confidence

HIGH — decoder is field-by-field over a fixed schema, tests cover all 5 required fields + null/array edge cases + multi-event ordering. No new deps, no new modules, every edit traces to the plan.

---
`/safer:review-senior` required before merge.

### CODE REVIEW

**Review by chughtapan — COMMENTED** [review]
## Verdict
request-changes (posted as comment — GitHub does not allow requesting changes on one's own PR)

## Scope
Claimed modality: implement-senior
Observed shape: tier=senior, files=2, modules=1 production + 1 test, no new module, no new deps
Public surface added: `DecodeError`, `DecodeErrorSink` (plan-named), `MoltzapRegistrar` callback widened to `unknown`, `register()` gains `diag` param — all plan-named per sbd#108.
Match: yes

## Craft findings

**[P1 — blocks merge] `v2/moltzap/listener.ts:33` — `field: string` in exported `DecodeError`**

`DecodeError` is re-exported from `v2/moltzap/index.ts`. The decoder only ever assigns six values to `field`: `"."`, `"messageId"`, `"conversationId"`, `"senderId"`, `"bodyText"`, `"receivedAtMs"`. Typing it `string` means any consumer that switches over `field` gets no exhaustiveness help from the type system:

```ts
switch (e.field) {
  case ".": ...
  case "messageId": ...
  // forgot receivedAtMs — TypeScript cannot flag this
  default:
}
```

Principle 1: constraints belong in the type, not in downstream guard code.

Fix: narrow `field` to the exact union:
```ts
readonly field: "." | "messageId" | "conversationId" | "senderId" | "bodyText" | "receivedAtMs";
```

Same narrowing applies to `checkField`'s `field` parameter (line 120) and `type` parameter (line 121 — only `"string"` and `"number"` are ever passed; the literal union catches a misspelling at compile time).

**[Nit — consistency] `v2/moltzap/listener.ts:110-112` — branding constructors bypassed in decoder**

`types.ts` exports `asMoltzapMessageId`, `asMoltzapConversationId`, `asMoltzapSenderId` as the canonical branding point. The decoder does direct `as MoltzapMessageId` casts. Doesn't affect correctness (the `checkField` validation ensures the value is a string before the cast), but branding should route through the constructor, not scattered `as` casts. Fix: import the three functions as values (currently only their types are imported) and use `asMoltzapMessageId(r["messageId"] as string)` etc.

## Scope alignment

| AC | Addressed | Evidence |
|---|---|---|
| Schema validation at listener boundary | ✓ | `wrappedCb` calls `decodeMoltzapInbound` before `cb` |
| Typed error, dropped + diagnostic, no throw | ✓ | `DecodeError` type; `diag(decoded.error); return;`; no `throw` in decode path |
| Unit tests for valid/invalid/edge shapes | ✓ | 9 new tests across 2 describe blocks |
| No buffering (I7) | ✓ | drop-and-return, no queue |

All three acceptance criteria met. No scope creep.

## Tests

Strong. Four valid-event tests (including `messageId: ""` and `receivedAtMs: 0` edge cases), five invalid-event tests (null, missing field, wrong type, array, multi-event ordering). The multi-event test is particularly good — verifies per-event independent decode without contamination.

Minor gap: only `messageId` is tested for missing field; `bodyText` and `receivedAtMs` are not individually covered. Given the shared `checkField` helper this is defensible — noting for completeness, not blocking.

## Notes

The `??`-chained `checkField` returns the first failure only — correct per spec (no accumulation required). A comment confirming the single-error-on-first-failure intent would prevent a future reader from "fixing" it into a multi-error accumulator.

File: v2/moltzap/listener.ts [hunk 1]
```diff
@@ -4,60 +4,121 @@
...
+// ── Decode error ─────────────────────────────────────────────────────
+//
+// sbd#108 follow-up: "typed error for malformed events". Named per field
+// so the diagnostic sink can surface exactly which part of the raw SDK
+// event was wrong without leaking large payloads.
+
+export type DecodeError = {
+  readonly _tag: "DecodeError";
+  /** The field path that failed validation ("." for non-object input). */
+  readonly field: string;
```
Comment (chughtapan) [inline v2/moltzap/listener.ts]: **P1 — blocks merge.** `field` is typed `string`, but the decoder only ever produces six values: `"."`, `"messageId"`, `"conversationId"`, `"senderId"`, `"bodyText"`, `"receivedAtMs"`. This type is re-exported from the barrel. A consumer switching over `error.field` gets zero exhaustiveness protection. Narrow to the literal union.

File: v2/moltzap/listener.ts [hunk 2]
```diff
@@ -4,60 +4,121 @@
...
+    senderId: r["senderId"] as MoltzapSenderId,
+    bodyText: r["bodyText"] as string,
+    receivedAtMs: r["receivedAtMs"] as number,
+  });
+}
+
+function checkField(
+  r: Record<string, unknown>,
+  field: string,
+  type: string,
```
Comment (chughtapan) [inline v2/moltzap/listener.ts]: **P1 — nit.** Only `"string"` and `"number"` are ever passed here. Change to `type: "string" | "number"`. A misspelling like `"nubmer"` silently passes today; the literal union catches it at compile time.

File: v2/moltzap/listener.ts [hunk 3]
```diff
@@ -4,60 +4,121 @@
...
+  const r = raw as Record<string, unknown>;
+  const fieldErr =
+    checkField(r, "messageId", "string") ??
+    checkField(r, "conversationId", "string") ??
+    checkField(r, "senderId", "string") ??
+    checkField(r, "bodyText", "string") ??
+    checkField(r, "receivedAtMs", "number");
+  if (fieldErr) return err(fieldErr);
+  return ok({
+    messageId: r["messageId"] as MoltzapMessageId,
```
Comment (chughtapan) [inline v2/moltzap/listener.ts]: **Nit — consistency.** `asMoltzapMessageId`, `asMoltzapConversationId`, `asMoltzapSenderId` are the canonical branding constructors in `types.ts`. Use them here instead of raw `as` casts. Call would be `asMoltzapMessageId(r["messageId"] as string)`. Requires importing the three functions as values (currently only their types are imported).

**Comment by chughtapan** [comment]
Review fixes applied (d3f00f6):

- **P1**: `DecodeError.field` narrowed to `"." | "messageId" | "conversationId" | "senderId" | "bodyText" | "receivedAtMs"`. `checkField` params narrowed: `field` → same literal union (sans `"."`), `type` → `"string" | "number"`.
- **Nit**: `as MoltzapMessageId/ConversationId/SenderId` casts replaced with `asMoltzapMessageId/asMoltzapConversationId/asMoltzapSenderId` constructors; imported as values.
- **Comment**: single-error-on-first-failure intent of `??`-chain noted inline.

All 174 tests pass.

**Review by chughtapan — COMMENTED** [review]
## Verdict
APPROVE (comment — cannot self-approve on GitHub)

## Scope
Claimed modality: impl-senior
Observed shape: 2 files, 1 module (v2/moltzap), 0 new deps, plan-named public exports only
Match: yes — DecodeError, DecodeErrorSink named in sbd#108 follow-up; no new module, no new deps

## R1 blockers resolved

- **listener.ts:38** `DecodeError.field` narrowed to `"." | "messageId" | "conversationId" | "senderId" | "bodyText" | "receivedAtMs"` ✓
- **listener.ts:104–107** `checkField` `field` param narrowed to the 5-field union; `type` param narrowed to `"string" | "number"` ✓
- **listener.ts:115–118** `asMoltzapMessageId`, `asMoltzapConversationId`, `asMoltzapSenderId` used — not raw `as` casts ✓
- **listener.ts:97** `??` short-circuit comment present ✓

## Craft findings
None.

P1: `DecodeError.field` and `checkField` params are literal unions. Post-validation `as string`/`as number` casts are safe (type already verified by `checkField`).
P2: Boundary owned by this module — `MoltzapRegistrar` receives `(event: unknown) => void`; all 5 required fields validated before bridge sees the event.
P3: No throw in `wrappedCb`; catch returns typed `SDKRejected`; `DecodeError` surfaces through injected `diag` sink.
P4: No new switches; decoder terminates in explicit `return ok(...)`.

## Scope alignment (#132 acceptance criteria)
- Schema validation at listener boundary → `wrappedCb` + `decodeMoltzapInbound` ✓
- Typed error, dropped + diagnostic, no throw → `DecodeError` + `diag` sink ✓
- Unit tests valid/invalid/edge → 4 valid + 5 invalid = 9 new tests ✓

## Tests
Comprehensive. Covers null, missing field, wrong type, array, multi-event ordering, empty-string messageId, receivedAtMs=0. All 7 pre-existing tests updated. Nothing gutted.

## Notes
`bodyText` and `receivedAtMs` correctly omit branding constructors — they are unbranded in `MoltzapInbound`. No issue.

**Ready to merge.**
